src/Conversion/ONNXToTOSA/Tensor/Flatten.cpp

```diff
@@ -0,0 +1,83 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//===---------------- Flatten.cpp - Flatten Op --------------------===//
+//
+// Copyright (c) 2022 Advanced Micro Devices, Inc.
+//
+// =============================================================================
+//
+// This file lowers ONNX ReshapeOp to TOSA dialect.
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Dialect/Tosa/IR/TosaOps.h"
+#include "mlir/Interfaces/InferTypeOpInterface.h"
+#include "src/Conversion/ONNXToTOSA/ONNXToTOSACommon.hpp"
+#include "src/Conversion/ONNXToTOSA/ONNXToTOSALegalizeUtils.hpp"
+#include "src/Dialect/ONNX/ONNXOps.hpp"
+#include "llvm/ADT/SmallVector.h"
+
+using namespace mlir;
+
+namespace onnx_mlir {
+
+namespace {
+
+class ONNXFlattenLoweringToTOSA : public OpConversionPattern<ONNXFlattenOp> {
+public:
+  using OpConversionPattern::OpConversionPattern;
+  using OpAdaptor = typename ONNXFlattenOp::Adaptor;
+  LogicalResult matchAndRewrite(ONNXFlattenOp op, OpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+
+    Value input = adaptor.input();
+    int64_t axis = adaptor.axis();
+    auto inputType = input.getType().cast<ShapedType>();
+
+    auto resultType = op.getResult().getType();
+
+    if (axis < 0) {
+      axis += inputType.getRank();
+    }
+
+    llvm::SmallVector<int64_t> newShape;
+    if (axis == 0) {
+      newShape.push_back(1);
+      int64_t lastShape = 1;
+      for (const auto &axis : inputType.getShape()) {
+        lastShape *= axis;
+      }
+      newShape.push_back(lastShape);
+    } else {
+      auto inputShape = inputType.getShape();
+      int64_t firstShape = 1;
+      for (int i = 0; i < axis; i++) {
+        firstShape *= inputShape[i];
+      }
+      newShape.push_back(firstShape);
+      int64_t secondShape = 1;
+      for (int i = axis; i < inputType.getRank(); i++) {
+        secondShape *= inputShape[i];
+      }
+      newShape.push_back(secondShape);
+    }
+    ArrayAttr shapeAttr = rewriter.getI64ArrayAttr(newShape);
+
+    tosa::CreateReplaceOpAndInfer<tosa::ReshapeOp>(
+        rewriter, op, resultType, input, shapeAttr);
```

Comment:
You probably don't need to run shape inference for this operator as the shapes should be statically known, or is there any reason to do so?

Reply:
My understanding is that if onnx only has dynamic shapes, using the infer functions eliminates the need to run a shape inference pass. But I am not sure on this. Maybe upstream people can clarify this?

Reply:
That's actually something I was wondering too. Though, if ONNX only has dynamic shapes, what happens to the new shape that you are trying to create? How are the unknown dimensions represented, and can you multiply them as you are doing?

Reply:
The behavior is really weird. The shape calculations are all correct, but because the test output should be a dynamic shape, the original result remaines live and the test fails

Reply:
I think we should wait for feedback from ARM about what their intention was with this infering of shapes and when not to use it

Reply:
I'd ask @ljfitz before going to the arm people though.
But for us, it should not be a problem because we run shape inference far earlier in the flow, so you can resolve this comment for the moment and create a followup PR if changes are needed imo